pkg/gateway/multiparts/tracker.go

```diff
@@ -38,82 +46,70 @@ var (
 	ErrInvalidMetadataSrcFormat = errors.New("invalid metadata source format")
 )
 
-func NewTracker(adb db.Database) Tracker {
+func NewTracker(ms kv.StoreMessage) Tracker {
 	return &tracker{
-		db: adb,
+		store: ms,
+	}
+}
+
+func multipartFromProto(pb *MultipartUploadData) *MultipartUpload {
+	return &MultipartUpload{
+		UploadID:        pb.UploadId,
+		Path:            pb.Path,
+		CreationDate:    pb.CreationDate.AsTime(),
+		PhysicalAddress: pb.PhysicalAddress,
+		Metadata:        pb.Metadata,
+		ContentType:     pb.ContentType,
+	}
+}
+
+func protoFromMultipart(m *MultipartUpload) *MultipartUploadData {
+	return &MultipartUploadData{
+		UploadId:        m.UploadID,
+		Path:            m.Path,
+		CreationDate:    timestamppb.New(m.CreationDate),
+		PhysicalAddress: m.PhysicalAddress,
+		Metadata:        m.Metadata,
+		ContentType:     m.ContentType,
 	}
 }
 
 func (m *tracker) Create(ctx context.Context, multipart MultipartUpload) error {
 	if multipart.UploadID == "" {
 		return ErrInvalidUploadID
 	}
-	_, err := m.db.Transact(ctx, func(tx db.Tx) (interface{}, error) {
-		_, err := tx.Exec(`INSERT INTO gateway_multiparts (upload_id,path,creation_date,physical_address, metadata, content_type)
-			VALUES ($1, $2, $3, $4, $5, $6)`,
-			multipart.UploadID, multipart.Path, multipart.CreationDate, multipart.PhysicalAddress, multipart.Metadata, multipart.ContentType)
-		return nil, err
-	})
+	_, err := m.Get(ctx, multipart.UploadID)
+	if err == nil {
+		return ErrAlreadyExists
+	}
+	path := kv.NewPath(multipartPrefix, multipart.UploadID)
+	err = m.store.SetMsg(ctx, path, protoFromMultipart(&multipart))
```

Comment:
@nopcoder - If we use SetIf in this instance - we will get an ambiguous error on ErrPredicateFailed which we will need to interpret as ErrAlreadyExists. We will also not save on any API calls since SetIf uses Get as well

Reply:
Understood about the error type migration - but it is in place to handle the 'transaction' part of setting a new value, just in case it doesn't exist.

Reply:
I've modified the SetIf function in store_message. Please have another look